### PR TITLE
KAFKA-5213; Mark a MemoryRecordsBuilder as full as soon as the append stream is closed

### PR DESCRIPTION
**Pull request by apurvam**

### CODE REVIEW

**Comment by apurvam** [comment]
cc @hachikuji @ijuma

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/ProducerBatchTest.java [hunk 1]
```diff
@@ -70,4 +73,14 @@ public void testLargeFullOldNowExpire() {
...
         assertFalse(batch.maybeExpire(10240, 10240L, now - 2L, 10240L, true));
     }
+
+    @Test
+    public void testShouldNotAppendOnceRecordBatchGetsFull() {
+        ProducerBatch batch = new ProducerBatch(new TopicPartition("topic", 1), memoryRecordsBuilder, now);
+        FutureRecordMetadata result0 = batch.tryAppend(now, null, new byte[10], Record.EMPTY_HEADERS, null, now);
+        assertNotNull(result0);
+        memoryRecordsBuilder.closeForRecordAppends();
+        assertFalse(memoryRecordsBuilder.hasRoomFor(now, null, new byte[10]));
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/ProducerBatchTest.java]: I'd add this same test (but `assertTrue`) just before `closeForRecordAppends` to verify that the `closeForRecordApends` is what causes the change.

Comment (apurvam) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/ProducerBatchTest.java]: Good suggestion, updated.

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java [hunk 2]
```diff
@@ -664,7 +664,7 @@ public boolean isClosed() {
     public boolean isFull() {
         // note that the write limit is respected only after the first record is added which ensures we can always
         // create non-empty batches (this is used to disable batching when the producer's batch size is set to 0).
-        return isClosed() || (this.numRecords > 0 && this.writeLimit <= estimatedBytesWritten());
+        return appendStreamIsClosed || (this.numRecords > 0 && this.writeLimit <= estimatedBytesWritten());
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:667]: Is it possible that `isClosed()` is true, but `appendStreamIsClosed` is false?

Comment (apurvam) [reply clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:667]: No. `MemoryRecordsBuilder.close` calls `closeForRecordAppends` after performing validations.

**Review by ijuma — APPROVED** [review]
LGTM, thanks. @hachikuji, do you want to take a look?

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the quick fix!

**Comment by ijuma** [comment]
Bonus points for including the test from the start. :)

**Pull request closed without merging**